File: code/tcp-udp/TcpServer.java
import java.io.*;
import java.net.*;

/**
 * TCP 服务器示例
 * 演示 TCP 三次握手、数据传输过程
 *
 * 运行方式:
 *   javac TcpServer.java && java TcpServer
 */
public class TcpServer {

    public static void main(String[] args) throws IOException {
        int port = 8080;
        ServerSocket serverSocket = new ServerSocket(port);
        System.out.println("TCP Server 启动, 监听端口: " + port);
        System.out.println("等待客户端连接... (3 次握手)");

        try (Socket socket = serverSocket.accept();
             BufferedReader in = new BufferedReader(
                     new InputStreamReader(socket.getInputStream()));
             PrintWriter out = new PrintWriter(
                     socket.getOutputStream(), true)) {

            System.out.println("客户端已连接: " + socket.getInetAddress());
            System.out.println("--- 3 次握手完成 ---");

            String line;
            while ((line = in.readLine()) != null) {
                System.out.println("收到: " + line);
                out.println("ECHO: " + line);
            }

            System.out.println("客户端断开连接");
            System.out.println("--- 4 次挥手完成 ---");
        }
    }
}

File: code/tcp-udp/tcp_server.py
import socket


def main():
    host = '127.0.0.1'
    port = 8080

    server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    server.bind((host, port))
    server.listen(1)

    print(f'TCP Server 启动, 监听端口: {port}')
    print('等待客户端连接... (3 次握手)')

    conn, addr = server.accept()
    print(f'客户端已连接: {addr[0]}:{addr[1]}')
    print('--- 3 次握手完成 ---')

    with conn:
        while True:
            data = conn.recv(1024)
            if not data:
                break
            msg = data.decode()
            print(f'收到: {msg}')
            conn.sendall(f'ECHO: {msg}'.encode())

    print('客户端断开连接')
    print('--- 4 次挥手完成 ---')
    server.close()


if __name__ == '__main__':
    main()
